### PR TITLE
WIP: Adding typo correction.

### LINKED ISSUE #358: Typo correction
The following experience would be much better when there is a typo

```
λ t-rex --versoin
Do you mean "--version"? <-------------------------with a suggestion for typo
Unrecognized command or argument '--versoin'

t-rex:
  A command line testing tool for .NET

Usage:
  t-rex [options]

Options:
  --file                     .trx file(s) to parse
  --filter                   Only look at tests containing the specified text. "*" can be used as a wildcard.
  --format                   The format for the output. (Summary, JSON)
  --path                     Directory or directories to search for .trx files. Only the most recent .trx file in a given directory is
                             used.
  --hide-test-output         For failed tests, hide detailed test output. (Defaults to false.)
  --ansi-mode                Shows output formatted using ANSI sequences
  --virtual-terminal-mode    Shows output formatted using ANSI sequences
  --version                  Display version information
```

### PR DESCRIPTION
**Pull request by Keboo**
Seemed like a fun little thing to mock out. 
Opening this PR just to get feedback on if people want this finished or not.

Fixes #358

### CODE REVIEW

File: src/System.CommandLine/Invocation/InvocationExtensions.cs [hunk 1]
```diff
@@ -184,6 +186,24 @@ public static CommandLineBuilder CancelOnProcessTermination(this CommandLineBuil
             return builder;
         }
 
+        public static CommandLineBuilder UseTypoCorrections(
+            this CommandLineBuilder builder, int maxLevenshteinDistance = 3)
```
Comment (jeredm) [inline src/System.CommandLine/Invocation/InvocationExtensions.cs]: @Keboo Where did the default of 3 come from? I'm not saying it is a bad default. I am just curious.

Comment (Keboo) [reply src/System.CommandLine/Invocation/InvocationExtensions.cs]: Mostly from [here](https://www.xkcd.com/221/) and the simplest thing to get working. There is probably some better heuristics to use rather than just picking a number out of a hat.
I considered using something like:
```C#
string unmatchedToken = ...;
int maxLevenshteinDistance = (int)Math.Ceiling(unmatchedToken.Length / 2.0);
```
Open to other suggestions too.

The other part I am not completely happy with in single character aliases. You can imaging have a command with options `-a`, `-b`, and `-c` and then the user types `-d`. All of those would show as suggestions since they all have the same distance.

File: src/System.CommandLine/Invocation/TypoCorrection.cs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        public TypoCorrection(int maxLevenshteinDistance)
+        {
+            _maxLevenshteinDistance = maxLevenshteinDistance;
+        }
+
+        public void ProvideSuggestions(ParseResult result, IConsole console)
+        {
+            foreach (var token in result.UnmatchedTokens)
+            {
+                string suggestions = string.Join(", or ", GetPossibleTokens(result.CommandResult.Command, token).Select(x => $"'{x}'"));
```
[the remark this answers is not in the document]
Comment (Keboo) [reply src/System.CommandLine/Invocation/TypoCorrection.cs]: Completely agree. The message also needs to be moved out so it can be localized.

File: src/System.CommandLine/Invocation/TypoCorrection.cs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+namespace System.CommandLine.Invocation
+{
+    internal class TypoCorrection
+    {
+        private readonly int _maxLevenshteinDistance;
+
+        public TypoCorrection(int maxLevenshteinDistance)
+        {
+            _maxLevenshteinDistance = maxLevenshteinDistance;
```
Comment (colombod) [inline src/System.CommandLine/Invocation/TypoCorrection.cs]: should this check for >= 0 distance?

Comment (Keboo) [reply src/System.CommandLine/Invocation/TypoCorrection.cs]: Yes, assuming we keep this as an incoming parameter, that should be checked. Ideally it would be nice to eliminate the fixed value and use some heuristics based on the unmatched token.

File: src/System.CommandLine/Invocation/TypoCorrection.cs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        }
+
+        public void ProvideSuggestions(ParseResult result, IConsole console)
+        {
+            foreach (var token in result.UnmatchedTokens)
+            {
+                string suggestions = string.Join(", or ", GetPossibleTokens(result.CommandResult.Command, token).Select(x => $"'{x}'"));
+                console.Out.Write($"'{token}' was not matched, did you mean {suggestions}?");
```
Comment (jonsequitur) [inline src/System.CommandLine/Invocation/TypoCorrection.cs]: ...matched. Did you mean...

File: src/System.CommandLine/Invocation/InvocationExtensions.cs (hunk 1 above)
Comment (Keboo) [reply src/System.CommandLine/Invocation/InvocationExtensions.cs]: Certainly. I think there is an outstanding question of what is expected. I think for single character tokens it might be reasonable to either not provide any suggestions or do something different (perhaps only return suggestions for things that start with the same character as the token). I am not really sure what would be expected or desired.

Comment (jonsequitur) [reply src/System.CommandLine/Invocation/InvocationExtensions.cs]: I wonder if there should be more consistency of behavior here with tab suggestions. There, we do substring matching, which is a different kind of fuzzy matching. If someone types a single letter and hits tab, they get back all of the symbols containing that letter. That could work for typo suggestions as well.

Similarly, if someone types `myapp --nfo⇥`, maybe suggestions should allow tabbing through the typo-corrected variants?

In other words, is typo correction just a different interaction model for the existing `ISuggestionSource` abstractions, which should be expanded to consider misspellings in addition to the existing substring matching?
